### PR TITLE
Add filters to the search

### PR DESCRIPTION
**Pull request by donpaul999**
The users can filter the recipes by the **difficulty** or by the **time of preparation**.

- This filters can be accessed only from the **/search** page, after a search by title or ingredients was made.
- The filter call is sent to the `SearchServlet.java` and with the help of a [query string](https://cloud.google.com/appengine/docs/standard/java/search/query_strings#queries_on_text_and_html_fields) an index search is made with the new **parameters**.
- **Difficulty** is a string from the set - `{easy, medium, hard}`
- **Preparation time** is an integer value that stores the number of minutes required for the recipe to be prepared. The user  filters the recipes that can be cooked in less than a certain amount of time, such as 30 minutes, 1 hour or 2 hours.

### CODE REVIEW

File: src/main/java/com/google/sps/servlets/SearchServlet.java [hunk 1]
```diff
@@ -114,29 +205,88 @@ private Index getIndex(String indexName) {
...
+      time != 0
+    ) {
+      searchString = searchString + " AND ";
+    }
+    if (time != 0) {
+      searchString = searchString + "prep_time <=" + String.valueOf(time);
+    }
+
+    return searchString;
+  }
```
Comment (donpaul999) [inline src/main/java/com/google/sps/servlets/SearchServlet.java]: This function should look better.

**Review by ramaroberto — COMMENTED** [review]
Please split this PR in two PRs

File: src/main/java/com/google/sps/servlets/Recipe.java [hunk 2]
```diff
@@ -93,14 +101,14 @@ public void setDifficulty(String givenDiff) {
...
-   * Set the cook time as a string, to contain both the time itself and the unit of measure
+   * Set the cook time as a string, to contain both the time itself and the unit of measure.
    */
   public void setPrepTime(String givenPTime) {
     prepTime = givenPTime;
   }
 
   /**
-   * Set the cook time as a string, to contain both the time itself and the unit of measure
+   * Set the cook time as a string, to contain both the time itself and the unit of measure.
```
Comment (ramaroberto) [inline src/main/java/com/google/sps/servlets/Recipe.java]: This should be always minutes!

File: src/main/java/com/google/sps/servlets/Recipe.java [hunk 3]
```diff
@@ -52,12 +53,19 @@ public String getAuthor() {
...
+   * Get the list of ingredients for the recipe.
    */
   public ArrayList getIngredients() {
     return ingredients;
   }
 
+  /**
+   * Get the list of matched ingredients with a search for the recipe, from the snippet fields.
+   */
+  public ArrayList<String> getIngredientsMatched() {
```
Comment (ramaroberto) [inline src/main/java/com/google/sps/servlets/Recipe.java]: getMatchedIngredients sounds more natural

File: src/main/java/com/google/sps/servlets/SearchServlet.java [hunk 4]
```diff
@@ -55,46 +55,121 @@
...
+    String difficulty = getParameter("difficulty", request);
+    String time = getParameter("time", request);
+    HashMap<String, String> difficultiesMap = createMapOfDifficulties();
+    HashMap<Integer, String> timeIntervalsMap = createMapOfTimeIntervals();
 
-    List<Recipe> recipesListToReturn = recipesMatching(
-      request,
-      response,
-      stringToMatch
+    request.setAttribute("dataSearched", stringToMatch);
```
Comment (ramaroberto) [inline src/main/java/com/google/sps/servlets/SearchServlet.java]: The name of this parameter is really inconsistent, in 10 lines you're naming it in 3 different ways:

searchterm, stringToMatch, dataSearched.

Why the change of names? It's still the search term in all cases.

File: src/main/java/com/google/sps/servlets/SearchServlet.java [hunk 5]
```diff
@@ -55,46 +55,121 @@
...
+  /**
+   * Creates a map of values for the dropdown select for difficulties.
+   */
+  private HashMap<Integer, String> createMapOfTimeIntervals() {
+    HashMap<Integer, String> options = new HashMap<Integer, String>();
+    options.put(120, "Less than 2h");
+    options.put(60, "Less than 1h");
+    options.put(30, "Less than 30 minutes");
+    return options;
+  }
```
Comment (ramaroberto) [inline src/main/java/com/google/sps/servlets/SearchServlet.java]: Hmm, I don't understand why you're doing this, this should be only be done in the frontend, you're setting and receiving values (not strings) here only.

Please lets talk about this via GVC after you fix the rest of the PR.

File: src/main/java/com/google/sps/servlets/SearchServlet.java [hunk 6]
```diff
@@ -55,46 +55,121 @@
...
   public void doGet(HttpServletRequest request, HttpServletResponse response)
-    throws ServletException, IOException {
-    String stringToMatch = request
-      .getParameter("searchterm")
-      .replaceAll("[^a-zA-Z0-9]+", " ")
-      .trim();
-    request.setAttribute("dataSearched", stringToMatch);
+    throws ServletException, IOException, NullPointerException {
+    List<Recipe> recipesListToReturn = new ArrayList<>();
+    Integer timeValue;
```
Comment (ramaroberto) [inline src/main/java/com/google/sps/servlets/SearchServlet.java]: This should be defined closer yo where you're using it. Same with recipesListToReturn.

File: src/main/java/com/google/sps/servlets/SearchServlet.java [hunk 7]
```diff
@@ -55,46 +55,121 @@
...
+
+    try {
+      recipesListToReturn =
+        recipesMatching(
+          request,
+          response,
+          stringToMatch,
+          difficulty,
+          timeValue
+        );
```
Comment (ramaroberto) [inline src/main/java/com/google/sps/servlets/SearchServlet.java]: This should be

```
List<Recipe> recipesListToReturn =
        recipesMatching(
          request,
          response,
          stringToMatch,
          difficulty,
          timeValue
        );
request.setAttribute("recipesList", recipesListToReturn);
```

File: src/main/java/com/google/sps/servlets/SearchServlet.java [hunk 8]
```diff
@@ -55,46 +55,121 @@
...
+    } catch (SearchQueryException e) {
+      request.getRequestDispatcher("/search.jsp").forward(request, response);
```
Comment (ramaroberto) [inline src/main/java/com/google/sps/servlets/SearchServlet.java]: You need to return an error code in this case.

File: src/main/java/com/google/sps/servlets/SearchServlet.java [hunk 9]
```diff
@@ -55,46 +55,121 @@
...
+    options.put(60, "Less than 1h");
+    options.put(30, "Less than 30 minutes");
+    return options;
+  }
+
+  /**
+   * Replace and return the string with no commas and no multiple consecutive or trailing spaces.
+   */
+  private String trimAndRemoveCommas(String stringToRemove) {
+    if (!StringUtils.isBlank(stringToRemove)) {
```
Comment (ramaroberto) [inline src/main/java/com/google/sps/servlets/SearchServlet.java]: Do you need this check? What happens if you just try to replace on an empty string? I think it shouldn't be a problem.

File: src/main/java/com/google/sps/servlets/SearchServlet.java [hunk 10]
```diff
@@ -55,46 +55,121 @@
...
+      stringToRemove = stringToRemove.replaceAll("[,]+", " ");
+      stringToRemove = stringToRemove.replaceAll("  +", " ");
```
Comment (ramaroberto) [inline src/main/java/com/google/sps/servlets/SearchServlet.java]: This could be simplyfied to

stringToRemove.replaceAll("(,\s?)+", " ");

https://regexr.com/3f4vo

File: src/main/java/com/google/sps/servlets/SearchServlet.java [hunk 11]
```diff
@@ -55,46 +55,121 @@
...
+      "difficultySearchedValue",
+      difficultiesMap.get(difficulty)
     );
+    request.setAttribute("difficultyList", difficultiesMap);
+    request.setAttribute("cookingTimeList", timeIntervalsMap);
 
+    stringToMatch = trimAndRemoveCommas(stringToMatch);
+    difficulty = trimAndRemoveCommas(difficulty);
+    try {
+      timeValue = Integer.valueOf(trimAndRemoveCommas(time));
```
Comment (ramaroberto) [inline src/main/java/com/google/sps/servlets/SearchServlet.java]: Why are you trimming here if these are coming from selects?

File: src/main/java/com/google/sps/servlets/SearchServlet.java [hunk 12]
```diff
@@ -55,46 +55,121 @@
...
+    options.put(120, "Less than 2h");
+    options.put(60, "Less than 1h");
+    options.put(30, "Less than 30 minutes");
+    return options;
+  }
+
+  /**
+   * Replace and return the string with no commas and no multiple consecutive or trailing spaces.
+   */
+  private String trimAndRemoveCommas(String stringToRemove) {
```
Comment (ramaroberto) [inline src/main/java/com/google/sps/servlets/SearchServlet.java]: Why the parameter is called stringToRemove? You're not removing the string. I think the function should be called "sanitize" and the parameter could be only "text" or "string".

File: src/main/java/com/google/sps/servlets/SearchServlet.java [hunk 13]
```diff
@@ -55,46 +55,121 @@
...
     HttpServletRequest request,
     HttpServletResponse response,
-    String stringToMatch
+    String stringToMatch,
+    String difficulty,
+    Integer time
   )
     throws ServletException, IOException {
     DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
     List<Recipe> recipesMatchingList = new ArrayList<>();
```
Comment (ramaroberto) [inline src/main/java/com/google/sps/servlets/SearchServlet.java]: matchingRecipes

File: src/main/java/com/google/sps/servlets/SearchServlet.java [hunk 14]
```diff
@@ -55,46 +55,121 @@
...
+        Entity recipeEntity = getRecipeEntityFromDocument(datastore, document);
+        ArrayList<String> ingredientsMatched = returnMatchedIngredients(
+          document
         );
 
-        Recipe recipe = buildRecipe(recipeEntity);
+        Recipe recipe = buildRecipe(recipeEntity, ingredientsMatched);
         recipesMatchingList.add(recipe);
       } catch (EntityNotFoundException e) {
         response.setStatus(505);
```
Comment (ramaroberto) [inline src/main/java/com/google/sps/servlets/SearchServlet.java]: You should return an empty list in this case.

File: src/main/java/com/google/sps/servlets/SearchServlet.java [hunk 15]
```diff
@@ -55,46 +55,121 @@
...
+    Query query = buildQuery(stringToMatch, difficulty, time);
     Results<ScoredDocument> results = index.search(query);
-    for (ScoredDocument entity : results) {
+    for (ScoredDocument document : results) {
       try {
-        Entity recipeEntity = getRecipeEntityFromDocumentEntity(
-          datastore,
-          entity
+        Entity recipeEntity = getRecipeEntityFromDocument(datastore, document);
+        ArrayList<String> ingredientsMatched = returnMatchedIngredients(
```
Comment (ramaroberto) [inline src/main/java/com/google/sps/servlets/SearchServlet.java]: matchingIngredients

File: src/main/java/com/google/sps/servlets/SearchServlet.java (hunk 13 above)
Comment (ramaroberto) [reply src/main/java/com/google/sps/servlets/SearchServlet.java]: Also the variable should be declared just before the for.

File: src/main/java/com/google/sps/servlets/SearchServlet.java [hunk 16]
```diff
@@ -105,6 +180,22 @@ public void doGet(HttpServletRequest request, HttpServletResponse response)
     return recipesMatchingList;
   }
 
+  /**
+   * Returns a list of matched ingredients snippet by the search query.
+   */
+  private ArrayList<String> returnMatchedIngredients(ScoredDocument document) {
```
Comment (ramaroberto) [inline src/main/java/com/google/sps/servlets/SearchServlet.java]: returnMatchingIngredients

Always write the things in present with the exception when the past is related to the status / data.

In this case is just related to the fact that the query was executed before, but it's clearer to reflect it as "now".

File: src/main/java/com/google/sps/servlets/SearchServlet.java [hunk 17]
```diff
@@ -105,6 +180,22 @@ public void doGet(HttpServletRequest request, HttpServletResponse response)
...
+    ArrayList<String> ingredientsMatched = new ArrayList<>();
```
Comment (ramaroberto) [inline src/main/java/com/google/sps/servlets/SearchServlet.java]: matchingIngredients

File: src/main/java/com/google/sps/servlets/SearchServlet.java [hunk 18]
```diff
@@ -105,6 +180,22 @@ public void doGet(HttpServletRequest request, HttpServletResponse response)
...
+
+    List<Field> listFields = document.getExpressions();
```
Comment (ramaroberto) [inline src/main/java/com/google/sps/servlets/SearchServlet.java]: fields

don't repeat the type name in the variable name, it's clear from the definition.

File: src/main/java/com/google/sps/servlets/SearchServlet.java [hunk 19]
```diff
@@ -105,6 +180,22 @@ public void doGet(HttpServletRequest request, HttpServletResponse response)
...
+    for (Field field : listFields) {
+      if (!StringUtils.isBlank(field.getHTML())) ingredientsMatched.add(
```
Comment (ramaroberto) [inline src/main/java/com/google/sps/servlets/SearchServlet.java]: Add curly braces to this if.

File: src/main/java/com/google/sps/servlets/SearchServlet.java [hunk 20]
```diff
@@ -114,29 +205,88 @@ private Index getIndex(String indexName) {
...
+  private Query buildQuery(
+    String stringToMatch,
+    String difficulty,
+    Integer time
+  ) {
     Cursor responseCursor = Cursor.newBuilder().build();
     QueryOptions options = QueryOptions
       .newBuilder()
-      .setLimit(60)
+      .setLimit(10)
```
Comment (ramaroberto) [inline src/main/java/com/google/sps/servlets/SearchServlet.java]: Define the 10 as a class constant.

File: src/main/java/com/google/sps/servlets/SearchServlet.java [hunk 21]
```diff
@@ -114,29 +205,88 @@ private Index getIndex(String indexName) {
...
+
+    if (!StringUtils.isBlank(stringToMatch)) {
+      searchString =
+        searchString +
+        "((title=" +
+        stringToMatch.replaceAll(" ", " AND ") +
+        ")";
+    }
+    if (!StringUtils.isBlank(stringToMatch)) {
+      searchString = searchString + " OR ";
```
Comment (ramaroberto) [inline src/main/java/com/google/sps/servlets/SearchServlet.java]: This should be inside the previous if, it's the same condition.

File: src/main/java/com/google/sps/servlets/SearchServlet.java [hunk 22]
```diff
@@ -114,29 +205,88 @@ private Index getIndex(String indexName) {
...
+        searchString +
+        "(ingredients=" +
+        stringToMatch.replaceAll(" ", " AND ") +
+        "))";
+    }
+    if (
+      !StringUtils.isBlank(stringToMatch) && !StringUtils.isBlank(difficulty)
+    ) {
+      searchString = searchString + " AND ";
+    }
```
Comment (ramaroberto) [inline src/main/java/com/google/sps/servlets/SearchServlet.java]: You can delete one condition here and nest this if in the previous one.

File: src/main/java/com/google/sps/servlets/SearchServlet.java [hunk 23]
```diff
@@ -114,29 +205,88 @@ private Index getIndex(String indexName) {
...
+
+    if (!StringUtils.isBlank(difficulty)) {
+      searchString = searchString + "difficulty=\"" + difficulty + "\"";
+    }
+
+    if (
+      !StringUtils.isBlank(difficulty) &&
+      time != 0 ||
+      !StringUtils.isBlank(stringToMatch) &&
+      time != 0
```
Comment (ramaroberto) [inline src/main/java/com/google/sps/servlets/SearchServlet.java]: (!StringUtils.isBlank(difficulty) || !StringUtils.isBlank(stringToMatch)) && time != 0

File: src/main/java/com/google/sps/servlets/SearchServlet.java [hunk 24]
```diff
@@ -114,29 +205,88 @@ private Index getIndex(String indexName) {
...
+    String searchString = createSearchString(stringToMatch, difficulty, time);
+    Query query = Query.newBuilder().setOptions(options).build(searchString);
 
     return query;
   }
 
+  /**
+   * This function creates and returns the string used inside the search query.
+   */
+  private String createSearchString(
```
Comment (ramaroberto) [inline src/main/java/com/google/sps/servlets/SearchServlet.java:191]: Besides the comments I've wrote, which are for you to learn how to simplify conditions, you don't need to do things like:

!StringUtils.isBlank(stringToMatch) or !StringUtils.isBlank(difficulty)

Instead, you could always check for

!StringUtils.isBlank(searchString)

Basically, you need to prepend and AND if there's some text in the query already.

File: src/main/webapp/one-off/databaseInit.java [hunk 25]
```diff
@@ -187,15 +187,16 @@ private Entity buildRecipeEntity(
       "Lorem quam dolor dapibus ante, sit amet pellentesque turpis lacus eu ipsum. Duis quis mi ut tortor interdum efficitur quis at mi. Pellentesque quis mauris vel ligula commodo scelerisque. In vulputate quam nisl, vel sagittis ipsum molestie quis. Suspendisse quis ipsum a sem aliquam euismod mattis sed metus.";
     Random rd = new Random();
     Double number = rd.nextDouble();
+    Double prep_time = 30.00;
```
Comment (ramaroberto) [inline src/main/webapp/one-off/databaseInit.java]: Why is this a double if in the backend it's an integer?

File: src/main/java/com/google/sps/servlets/SearchServlet.java (hunk 9 above)
Comment (donpaul999) [reply src/main/java/com/google/sps/servlets/SearchServlet.java:128]: No, I don't have to check if the string is blank.

File: src/main/java/com/google/sps/servlets/SearchServlet.java (hunk 11 above)
Comment (donpaul999) [reply src/main/java/com/google/sps/servlets/SearchServlet.java]: Because the user can insert trash values inside the request directly from the link.

File: src/main/java/com/google/sps/servlets/SearchServlet.java (hunk 16 above)
Comment (donpaul999) [reply src/main/java/com/google/sps/servlets/SearchServlet.java]: I understand but the ingredients matching was made in the past, when the search is called. I only get the field from the ScoredDocument.

File: src/main/webapp/one-off/databaseInit.java (hunk 25 above)
Comment (donpaul999) [reply src/main/webapp/one-off/databaseInit.java]: I got some errors when requesting the value from the entity, I manage to solve them by using Double. I'll change the value to int.

File: src/main/java/com/google/sps/servlets/SearchServlet.java (hunk 9 above)
Comment (donpaul999) [reply src/main/java/com/google/sps/servlets/SearchServlet.java]: If I don't check if the string is blank I run into a `NullPointerException`.

File: src/main/java/com/google/sps/servlets/QueryForSize.java [hunk 26]
```diff
@@ -0,0 +1,39 @@
...
+import com.google.appengine.api.datastore.KeyFactory;
+import com.google.appengine.api.datastore.KeyFactory;
+import com.google.appengine.api.datastore.KeyRange;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.Query.SortDirection;
+
+public class QueryForSize {
+
+  public int countRecipes() {
```
Comment (ramaroberto) [inline src/main/java/com/google/sps/servlets/QueryForSize.java]: This method can be constant.

File: src/main/java/com/google/sps/servlets/QueryForSize.java [hunk 27]
```diff
@@ -0,0 +1,39 @@
...
+import com.google.appengine.api.datastore.FetchOptions;
+import com.google.appengine.api.datastore.Key;
+import com.google.appengine.api.datastore.KeyFactory;
+import com.google.appengine.api.datastore.KeyFactory;
+import com.google.appengine.api.datastore.KeyRange;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.Query.SortDirection;
+
+public class QueryForSize {
```
Comment (ramaroberto) [inline src/main/java/com/google/sps/servlets/QueryForSize.java]: I'd call this class QueryUtils and save all the query-related functions here, this is too specific.

File: src/main/java/com/google/sps/servlets/SearchServlet.java [hunk 28]
```diff
@@ -44,9 +45,11 @@
 
 @WebServlet("/search")
 public class SearchServlet extends HttpServlet {
+  // The number of recipes to be displayed on the search list.
+  private static final int RECIPES_LIMIT = 10;
 
   /**
-   * Search and returns a list of first 10 recipes with the title matching the given parameter(searchterm).
+   * Search and returns a list of first 10 recipes with the title matching the given parameters(searchterm, difficulty & time).
```
Comment (ramaroberto) [inline src/main/java/com/google/sps/servlets/SearchServlet.java]: the given parameters: searchterm, difficulty, and time.

Please write the comments following the correct grammar rules!

File: src/main/java/com/google/sps/servlets/SearchServlet.java [hunk 29]
```diff
@@ -55,64 +58,116 @@
...
-    List<Recipe> recipesListToReturn = recipesMatching(
-      request,
-      response,
-      stringToMatch
+    throws ServletException, IOException, NullPointerException {
+    String searchTerm = getParameter("searchterm", request);
+    String difficulty = getParameter("difficulty", request);
+    String time = getParameter("time", request);
+
+    request.setAttribute("searchTerm", searchTerm);
```
Comment (ramaroberto) [inline src/main/java/com/google/sps/servlets/SearchServlet.java]: is it searchterm or searchTerm? (look at line 61)

File: src/main/java/com/google/sps/servlets/SearchServlet.java [hunk 30]
```diff
@@ -55,64 +58,116 @@
...
+    request.setAttribute("difficulty", difficulty);
+    request.setAttribute("prepTime", time);
+
+    searchTerm = sanitizeString(searchTerm);
+    difficulty = sanitizeString(difficulty);
+    request.setAttribute(
+      "difficulty",
+      difficulty
     );
```
Comment (ramaroberto) [inline src/main/java/com/google/sps/servlets/SearchServlet.java]: You're doing this already in line 67. Please do the sanitization in the same line (inside the setAttribute) if that's the right thing.

File: src/main/java/com/google/sps/servlets/SearchServlet.java [hunk 31]
```diff
@@ -55,64 +58,116 @@
...
+      timeValue = Integer.valueOf(sanitizeString(time));
+    } catch (NumberFormatException e) {
+      timeValue = 0;
+    }
+
+    try {
+      List<Recipe> recipesListToReturn = recipesMatching(
+        request,
+        response,
+        searchTerm,
```
Comment (ramaroberto) [inline src/main/java/com/google/sps/servlets/SearchServlet.java]: Please apply sanitize here and delete it from line 70.

File: src/main/java/com/google/sps/servlets/SearchServlet.java [hunk 32]
```diff
@@ -55,64 +58,116 @@
...
+
+  /**
+   * If the fields exists, returns the parameter from the request, otherwise it returns an empty string.
+   */
+  private String getParameter(String field, HttpServletRequest request) {
+    String parameter = new String();
+    try {
+      parameter = request.getParameter(field);
+    } catch (NullPointerException e) {
+      parameter = "";
```
Comment (ramaroberto) [inline src/main/java/com/google/sps/servlets/SearchServlet.java]: return ""

File: src/main/java/com/google/sps/servlets/SearchServlet.java [hunk 33]
```diff
@@ -55,64 +58,116 @@
...
+    return stringToSanitize;
+  }
+
+  /**
+   * If the fields exists, returns the parameter from the request, otherwise it returns an empty string.
+   */
+  private String getParameter(String field, HttpServletRequest request) {
+    String parameter = new String();
+    try {
+      parameter = request.getParameter(field);
```
Comment (ramaroberto) [inline src/main/java/com/google/sps/servlets/SearchServlet.java]: return request.getParameter(field);

File: src/main/java/com/google/sps/servlets/SearchServlet.java [hunk 34]
```diff
@@ -55,64 +58,116 @@
...
+      stringToSanitize = stringToSanitize.replaceAll("(,\\s?)+", " ");
+    }
+    return stringToSanitize;
+  }
+
+  /**
+   * If the fields exists, returns the parameter from the request, otherwise it returns an empty string.
+   */
+  private String getParameter(String field, HttpServletRequest request) {
+    String parameter = new String();
```
Comment (ramaroberto) [inline src/main/java/com/google/sps/servlets/SearchServlet.java]: Delete this

File: src/main/java/com/google/sps/servlets/SearchServlet.java [hunk 35]
```diff
@@ -55,64 +58,116 @@
...
+    try {
+      parameter = request.getParameter(field);
+    } catch (NullPointerException e) {
```
Comment (ramaroberto) [inline src/main/java/com/google/sps/servlets/SearchServlet.java]: Instead of catching a null pointer exception you could check with an if directly

if (field != null) {
return request.getParameter(field);
}

return "";

File: src/main/java/com/google/sps/servlets/SearchServlet.java [hunk 36]
```diff
@@ -55,64 +58,116 @@
...
-        Recipe recipe = buildRecipe(recipeEntity, datastore);
-        recipesMatchingList.add(recipe);
+        Recipe recipe = buildRecipe(recipeEntity, matchingIngredients);
+        matchingRecipes.add(recipe);
       } catch (EntityNotFoundException e) {
         response.setStatus(505);
+        request.setAttribute("recipesList", new ArrayList<>());
         request.getRequestDispatcher("/search.jsp").forward(request, response);
-        return new ArrayList<>();
+        return matchingRecipes;
```
Comment (ramaroberto) [inline src/main/java/com/google/sps/servlets/SearchServlet.java]: Leaving the same comment here for the second time (for some reason this was changed again!):

You should return an empty list in this case.

**Review by ramaroberto — COMMENTED** [review]
There's something weird going on with the changes, the one from the random PR are shown here!

That's why I'm not approving the PR.

File: src/main/java/com/google/sps/servlets/QueryForSize.java (hunk 27 above)
Comment (donpaul999) [reply src/main/java/com/google/sps/servlets/QueryForSize.java]: I removed the class, because I found a way to use both Query types on the same class.

File: src/main/java/com/google/sps/servlets/SearchServlet.java (hunk 30 above)
Comment (donpaul999) [reply src/main/java/com/google/sps/servlets/SearchServlet.java]: I forgot to remove the duplication of setAttribute function usage.

File: src/main/java/com/google/sps/servlets/SearchServlet.java (hunk 36 above)
Comment (donpaul999) [reply src/main/java/com/google/sps/servlets/SearchServlet.java]: I changed by mistake when merging the commits.

File: src/main/java/com/google/sps/servlets/SearchServlet.java [hunk 37]
```diff
@@ -55,64 +58,96 @@
...
+    request.setAttribute("searchterm", searchterm);
+    request.setAttribute("difficulty", difficulty);
+    request.setAttribute("prepTime", time);
+
+    Integer timeValue;
+    try {
+      timeValue = Integer.valueOf(sanitizeString(time));
+    } catch (NumberFormatException e) {
+      timeValue = 0;
+    }
```
Comment (ramaroberto) [inline src/main/java/com/google/sps/servlets/SearchServlet.java]: Please modularize this in a function called "parseInt".

File: src/main/java/com/google/sps/servlets/SearchServlet.java [hunk 38]
```diff
@@ -165,15 +245,16 @@ private Recipe buildRecipe(Entity recipeEntity, DatastoreService datastore) {
     recipe.setName(name);
     recipe.setImage(imgURL);
     recipe.setDescription(description);
-    recipe.setPrepTime(prep_time_int);
+    recipe.setMatchingIngredient(ingredientsMatching);
+	  recipe.setPrepTime(prep_time);
```
Comment (ramaroberto) [inline src/main/java/com/google/sps/servlets/SearchServlet.java]: Fix indentation.

**Review by ramaroberto — APPROVED** [review]

File: src/main/java/com/google/sps/servlets/Recipe.java [hunk 39]
```diff
@@ -24,12 +24,13 @@
   private String name;
   private String description;
   private String difficulty;
-  private String cookTime;
+  private Integer cookTime;
```
Comment (bessmarche) [inline src/main/java/com/google/sps/servlets/Recipe.java]: Cook time and prep time are now just one parameter so you do not need them both.

**Review by bessmarche — APPROVED** [review]
